Update Intro/common-tricks (#3)
OI-wiki/i18n: docs/intro/common-tricks.md

This page mainly shares some common tricks used in the competition.

## Use locality

Locality refers to a program's tendency to refer to data items that are adjacent to other recently referenced data items, or those that were recently referenced themselves. Locality is divided into temporal locality and spatial locality.

-   Eliminate inefficiencies in loops, such as when traversing strings:
    ```cpp
    /* clang-format off */
    for (int i = 0; i < strlen(s); ++i);
    // better:
    for (int i = 0, len = strlen(s); i < len; ++i);
    ```

> -   The loop unfolds. The number of operations on the critical path in the entire calculation can be reduced by using appropriate loop expansion.

Reviewer: ```suggestion
-   Loop unrolling. The number of operations on the critical path in the entire calculation can be reduced by using appropriate loop expansion.
```

术语

#endif
```

> `#ifdef` will check whether there is a corresponding identifier defined by `#define` in the program, if there is, it will execute the following content, `#ifndef`, on the other hand, will be executed without defining the corresponding identifier The following statement.

Reviewer: ```suggestion
`#ifdef` will check whether there is a corresponding identifier defined by `#define` in the program, if there is, it will keep the following content, `#ifndef`, on the other hand, will be kept without defining the corresponding identifier The following statement.
```

原文有无，这里用 执行 不太恰当，应该用 保留

What is "Double Compare"? Specifically, the correctness of the program is verified by comparing the output of the two programs. You can compare the output of your program with the output of other programs (brute force or other contestants') to judge whether your program is correct.

Of course, since the "Double Compare" process needs to be carried out many times, we need to automate it by batch processing.

Specifically, we need a data generator, and two programs to be compared with.

Run [Data Generator](../topic/problemsetting/#_17) once each time, write the generated data to the input file, make the two programs read the data through the redirection method, and write the output to the specified file , Use the `fc` command under Windows to compare files (`diff` command under Linux) to verify the correctness of the program.

If you find an error in the program, you can use the data you just generated for debugging.

The general framework of the double compare is as follows:

```cpp
#include <stdio.h>
#include <stdlib.h>
int main() {
  // For Windows
  // Do not use file input and output

>   // Of course, this program can also be rewritten into a batch form

Reviewer: ```suggestion
  // Of course, this program can also be rewritten into a batch file or a shellscript file
```

Reviewer: 批处理指 .bat 文件

  while (1) {
    system("gen > test.in");  // The data generator writes the generated data to the input file
    system("test1.exe < test.in > a.out");  // Get program 1 output
    system("test2.exe < test.in > b.out");  // Get program 2 output
    if (system("fc a.out b.out")) {
      // This line of statements compares input and output
      // When fc returns 0, it means that the output is consistent
      // otherwise there is a difference
      system("pause");  // Easy to check the difference
      return 0;
      // The input data has been stored in the test.in file and can be used directly for debugging
    }
  }
}
```

## <span id="mempool">Memory pool</span>

When we need to allocate memory dynamically, frequent use of new/malloc will occupy a lot of time and space, and even generate a lot of memory fragments to reduce the performance of the program, which may make the originally correct program TLE/MLE.

> In this case, we need to use the "memory pool" technique: before actually using the memory, first we apply for the allocation of a certain size of memory as a spare. When dynamic allocation is needed, we can directly allocate a block from the spare memory.

Reviewer: ```suggestion
In this case, we need to use the "memory pool" technique: before actually using the memory, we first allocate of a certain size of memory as a memory pool. When dynamic allocation is needed, we can directly fetch a block from it.
```

Of course, in most OI problems, we can calculate the maximum memory required in advance and apply for allocation at one time.

> For example, the code to dynamically allocate $32$ signed integer array:

Reviewer: ```suggestion
For example, the code to dynamically allocate 32-bit signed integer array:
```
